vms/platformvm/vm_regression_test.go

```diff
@@ -1442,3 +1448,758 @@ func TestRemovePermissionedValidatorDuringPendingToCurrentTransitionTracked(t *t
+		Signer:    signer.NewProofOfPossession(sk2),
+		StakeOuts: stakedOuts,
+		ValidatorRewardsOwner: &secp256k1fx.OutputOwners{
+			Locktime:  0,
+			Threshold: 1,
+			Addrs: []ids.ShortID{
+				addr,
+			},
+		},
+		DelegatorRewardsOwner: &secp256k1fx.OutputOwners{
+			Locktime:  0,
+			Threshold: 1,
+			Addrs: []ids.ShortID{
+				addr,
+			},
+		},
+		DelegationShares: reward.PercentDenominator,
+	}
+	primaryRestartTx, err := txs.NewSigned(uPrimaryRestartTx, txs.Codec, signers)
+	require.NoError(err)
+	require.NoError(uPrimaryRestartTx.SyntacticVerify(vm.ctx))
+
+	require.NoError(vm.Builder.AddUnverifiedTx(primaryRestartTx))
+	blk, err = vm.Builder.BuildBlock(context.Background())
+	require.NoError(err)
+	require.NoError(blk.Verify(context.Background()))
+	require.NoError(blk.Accept(context.Background()))
+	require.NoError(vm.SetPreference(context.Background(), vm.manager.LastAccepted()))
+
+	// move time ahead, promoting restarted primary validator to current
+	currentTime = primaryReStartTime
+	vm.clock.Set(currentTime)
+	vm.state.SetTimestamp(currentTime)
+
+	blk, err = vm.Builder.BuildBlock(context.Background())
+	require.NoError(err)
+	require.NoError(blk.Verify(context.Background()))
+	require.NoError(blk.Accept(context.Background()))
+	require.NoError(vm.SetPreference(context.Background(), vm.manager.LastAccepted()))
+
+	_, err = vm.state.GetCurrentValidator(constants.PrimaryNetworkID, nodeID)
+	require.NoError(err)
+
+	primaryRestartHeight, err := vm.GetCurrentHeight(context.Background())
+	require.NoError(err)
+
+	// Show that validators are rebuilt with the right BLS key
+	for height := primaryStartHeight; height < primaryEndHeight; height++ {
+		require.NoError(checkValidatorBlsKeyIsSet(
+			vm.State,
+			nodeID,
+			constants.PrimaryNetworkID,
+			height,
+			uPrimaryTx.Signer.Key()),
+		)
+	}
+	for height := primaryEndHeight; height < primaryRestartHeight; height++ {
+		require.ErrorIs(checkValidatorBlsKeyIsSet(
+			vm.State,
+			nodeID,
+			constants.PrimaryNetworkID,
+			primaryEndHeight,
+			uPrimaryTx.Signer.Key()),
+			database.ErrNotFound,
+		)
+	}
+	require.NoError(checkValidatorBlsKeyIsSet(
+		vm.State,
+		nodeID,
+		constants.PrimaryNetworkID,
+		primaryRestartHeight,
+		uPrimaryRestartTx.Signer.Key()),
+	)
+
+	for height := subnetStartHeight; height < subnetEndHeight; height++ {
+		require.NoError(checkValidatorBlsKeyIsSet(
+			vm.State,
+			nodeID,
+			subnetID,
+			height,
+			uPrimaryTx.Signer.Key()),
+		)
+	}
+
+	for height := subnetEndHeight; height <= primaryRestartHeight; height++ {
+		require.ErrorIs(checkValidatorBlsKeyIsSet(
+			vm.State,
+			nodeID,
+			subnetID,
+			primaryEndHeight,
+			uPrimaryTx.Signer.Key()),
+			database.ErrNotFound,
+		)
+	}
+}
+
+func Test_RegressionPrimaryNetworkValidatorEmptyBLSKeyDiff(t *testing.T) {
+	// A primary network validator has an empty BLS key. Then it restakes
+	// adding the BLS key. Querying the validator set back when BLS key was empty
+	// must return an empty BLS key.
```

Comment:
is it possible to merge the tests (possibly with some parameterization around what keys to expect at what heights and how to issue the staking TXs for the primary validator?)